### PR TITLE
Profile memory allocations of the CArena using TinyProfiler

### PR DESCRIPTION
**Pull request by AlexanderSinn**
## Summary

The following per-section exclusive metrics of all CArenas are collected:
* Nalloc
Number of memory allocations inside a profiler section.
* Nfree
How many of the allocations counted in Nalloc where freed before `Finalize`. Works if the memory was freed in a different section compared to where it was allocated.
* AvgMem
The amount of memory allocated in a section averaged over execution time, taking into account how long it takes until the memory is freed again.
* MaxMem
The maximum amount of memory allocated by a section at any given time. In contrast to the other metrics, MaxMem is not additive between two sections, only satisfying the triangle inequality.

If more the one rank is used, Nalloc and Nfree are summed between all ranks while for AvgMem and MaxMem the minimum, average and maximum value across all ranks is printed.

Sample output from HiPACE++ (1 Rank)
```
Device Memory Usage:
---------------------------------------------------------------
Name                            Nalloc  Nfree  AvgMem    MaxMem
---------------------------------------------------------------
The_Device_Arena::Initialize()       1      1   4   B  8192 KiB
---------------------------------------------------------------

Managed Memory Usage:
---------------------------------------------------------------------------
Name                                      Nalloc  Nfree    AvgMem    MaxMem
---------------------------------------------------------------------------
The_Arena::Initialize()                        1      1    43 KiB    29 GiB
sortBeamParticlesByBox()                      44     44  7377 MiB    11 GiB
BeamParticleContainer::InitParticles          10     10    17 MiB  7390 MiB
PlasmaParticleContainer::InitParticles        34     34  4554 MiB  4737 MiB
Fields::AllocData()                            3      3  2820 MiB  2820 MiB
Hipace::ExplicitMGSolveBxBy()                 45     45  1332 MiB  1366 MiB
DenseBins<T>::buildGPU                        36     36  1182 MiB  1210 MiB
FFTPoissonSolverDirichlet::define()            3      3   383 MiB   383 MiB
AnyDST::CreatePlan()                           2      2   255 MiB   255 MiB
ResizeRandomSeed                               1      1    40 MiB    40 MiB
AdaptiveTimeStep::Calculate()                  4      4     6   B   972 KiB
hpmg::MultiGrid::solve1()                  17320  17320   108 KiB   486 KiB
Hipace::InitData()                             7      7   273   B   243 KiB
MultiPlasma::InitData()                        2      2     0   B   243 KiB
main()                                         3      3    79   B    80   B
Fields::Copy()                                 1      1    62   B    64   B
DepositCurrent_PlasmaParticleContainer()   10000  10000     1   B    16   B
---------------------------------------------------------------------------

Pinned Memory Usage:
-----------------------------------------------------------------
Name                            Nalloc  Nfree    AvgMem    MaxMem
-----------------------------------------------------------------
Hipace::Evolve()                     1      1  1682 MiB  1718 MiB
The_Pinned_Arena::Initialize()       1      1    52   B  8192 KiB
Hipace::ExplicitMGSolveBxBy()        1      1   998   B  1024   B
sortBeamParticlesByBox()            24     24   127   B   256   B
Hipace::InitData()                  37     37   207   B   240   B
main()                              41     41    79   B    96   B
Fields::Copy()                       1      1    62   B    64   B
AdaptiveTimeStep::Calculate()        4      4     0   B    32   B
MultiPlasma::InitData()              2      2     0   B    16   B
hpmg::MultiGrid::solve1()        17320  17320     6   B    16   B
-----------------------------------------------------------------

Cpu (amrex_mempool) Memory Usage:
-----------------------------------------------------
Name                  Nalloc  Nfree  AvgMem    MaxMem
-----------------------------------------------------
amrex_mempool_init()       1      1  25   B  8192 KiB
-----------------------------------------------------
```
4 Ranks:
```
Device Memory Usage:
---------------------------------------------------------------------------------------------------------------------
Name                            Nalloc  Nfree  AvgMem min  AvgMem avg  AvgMem max  MaxMem min  MaxMem avg  MaxMem max
---------------------------------------------------------------------------------------------------------------------
The_Device_Arena::Initialize()       4      4       4   B       4   B       5   B    8192 KiB    8192 KiB    8192 KiB
---------------------------------------------------------------------------------------------------------------------

Managed Memory Usage:
--------------------------------------------------------------------------------------------------------------------------------
Name                                      Nalloc   Nfree  AvgMem min  AvgMem avg  AvgMem max  MaxMem min  MaxMem avg  MaxMem max
--------------------------------------------------------------------------------------------------------------------------------
The_Arena::Initialize()                        4       4      16 KiB      19 KiB      23 KiB      29 GiB      29 GiB      29 GiB
sortBeamParticlesByBox()                     657     657    1458 MiB    1915 MiB    2808 MiB    4221 MiB    5987 MiB      11 GiB
MultiBeam::PackLocalGhostParticles()          10      10       0   B     500 MiB    2003 MiB       0   B    2078 MiB    8313 MiB
BeamParticleContainer::InitParticles          10      10       0   B    1252 KiB    5010 KiB       0   B    1847 MiB    7390 MiB
PlasmaParticleContainer::InitParticles       136     136    4556 MiB    4557 MiB    4557 MiB    4737 MiB    4737 MiB    4737 MiB
Hipace::Wait()                               301     301    1836 KiB    4227 KiB    5413 KiB    3695 MiB    3695 MiB    3695 MiB
Fields::AllocData()                           12      12    2820 MiB    2820 MiB    2820 MiB    2820 MiB    2820 MiB    2820 MiB
Hipace::ExplicitMGSolveBxBy()                180     180    1011 MiB    1176 MiB    1363 MiB    1366 MiB    1366 MiB    1366 MiB
DenseBins<T>::buildGPU                       744     744     240 MiB     283 MiB     325 MiB     605 MiB     605 MiB     605 MiB
FFTPoissonSolverDirichlet::define()           12      12     383 MiB     383 MiB     383 MiB     383 MiB     383 MiB     383 MiB
AnyDST::CreatePlan()                           8       8     255 MiB     255 MiB     255 MiB     255 MiB     255 MiB     255 MiB
ResizeRandomSeed                               4       4      40 MiB      40 MiB      40 MiB      40 MiB      40 MiB      40 MiB
AdaptiveTimeStep::Calculate()                 66      66       2   B       3   B       5   B     972 KiB     972 KiB     972 KiB
hpmg::MultiGrid::solve1()                 154398  154398      86 KiB      99 KiB     111 KiB     486 KiB     486 KiB     486 KiB
Hipace::InitData()                            28      28     271   B     271   B     272   B     243 KiB     243 KiB     243 KiB
MultiPlasma::InitData()                        8       8       0   B       0   B       0   B     243 KiB     243 KiB     243 KiB
Hipace::Notify()                             196     196       0   B       0   B       2   B     128   B     128   B     128   B
main()                                        12      12      79   B      79   B      79   B      80   B      80   B      80   B
Fields::Copy()                                 4       4      47   B      54   B      63   B      64   B      64   B      64   B
DepositCurrent_PlasmaParticleContainer()   80000   80000       1   B       1   B       1   B      16   B      16   B      16   B
--------------------------------------------------------------------------------------------------------------------------------

Pinned Memory Usage:
----------------------------------------------------------------------------------------------------------------------
Name                            Nalloc   Nfree  AvgMem min  AvgMem avg  AvgMem max  MaxMem min  MaxMem avg  MaxMem max
----------------------------------------------------------------------------------------------------------------------
Hipace::Notify()                   245     245     554 MiB     932 MiB    1166 MiB    3051 MiB    3051 MiB    3051 MiB
Hipace::Wait()                     245     245       9 MiB      20 MiB      32 MiB    3051 MiB    3051 MiB    3051 MiB
Hipace::Evolve()                     4       4     317 MiB     369 MiB     428 MiB     429 MiB     429 MiB     429 MiB
The_Pinned_Arena::Initialize()       4       4      21   B      24   B      27   B    8192 KiB    8192 KiB    8192 KiB
Hipace::ExplicitMGSolveBxBy()        4       4     757   B     881   B    1022   B    1024   B    1024   B    1024   B
sortBeamParticlesByBox()           576     576     287   B     287   B     287   B     416   B     416   B     416   B
Hipace::InitData()                 148     148     207   B     207   B     207   B     240   B     240   B     240   B
main()                             164     164      79   B      79   B      79   B      96   B      96   B      96   B
Fields::Copy()                       4       4      47   B      54   B      63   B      64   B      64   B      64   B
AdaptiveTimeStep::Calculate()       66      66       0   B       0   B       0   B      32   B      32   B      32   B
MultiPlasma::InitData()              8       8       0   B       0   B       0   B      16   B      16   B      16   B
hpmg::MultiGrid::solve1()       154398  154398       5   B       5   B       7   B      16   B      16   B      16   B
----------------------------------------------------------------------------------------------------------------------

Cpu (amrex_mempool) Memory Usage:
-----------------------------------------------------------------------------------------------------------
Name                  Nalloc  Nfree  AvgMem min  AvgMem avg  AvgMem max  MaxMem min  MaxMem avg  MaxMem max
-----------------------------------------------------------------------------------------------------------
amrex_mempool_init()       4      4       9   B       9   B      11   B    8192 KiB    8192 KiB    8192 KiB
-----------------------------------------------------------------------------------------------------------
```

## Additional background

A major difficulty in implementing this is that different omp threads could be in different functions. This means that each thread needs its own profiler function stack. Additionally multiple threads may share the same TinyProfiler object.

To clean up the Profiler output, profiler sections were added to arena preallocate/initialize functions. Additionally `amrex::DeallocateRandomSeedDevArray()` was moved before `BL_TINY_PROFILE_FINALIZE()` so the deallocation is not missing in the profile.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

File: Src/Base/AMReX_CArena.cpp [hunk 1]
```diff
@@ -4,6 +4,12 @@
 #include <AMReX_Gpu.H>
 #include <AMReX_ParallelReduce.H>
 
+#ifdef AMREX_TINY_PROFILING
+#include <AMReX_TinyProfiler.H>
+#else
+struct MemStat {};
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_CArena.cpp]: Could we add `namespace amrex` here?

File: Src/Base/AMReX_TinyProfiler.H [hunk 2]
```diff
@@ -84,20 +112,42 @@ private:
...
     bool uCUPTI;
+    bool in_parallel_region;
     int global_depth;
     std::vector<Stats*> stats;
 
+
+    static std::deque<const TinyProfiler*> mem_stack;
+#ifdef AMREX_USE_OMP
+    static std::deque<const TinyProfiler*> mem_stack_thread_private;
+#pragma omp threadprivate(mem_stack_thread_private)
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_TinyProfiler.H]: It appears that some compilers do not work well with omp threadprivate.  In `AMReX_BLBackTrace.H`, we have to do something like this.
```
struct BLBackTrace
{   
    ...
    static std::stack<std::pair<std::string, std::string> > bt_stack;
// threadprivate here doesn't work with Cray, Intel, and Fujitsu                                                                                                             
#if defined(AMREX_USE_OMP) && !defined(_CRAYC) && !defined(__INTEL_COMPILER) && !defined(__PGI) && !defined(__NVCOMPILER) && !defined(__FUJITSU)
#pragma omp threadprivate(bt_stack)
#endif
};
```
This is a hack just to get the code to compile.  We should fix it in a better way.

For the threadprivate variable here, we could make it a vector and resize it to the number of max threads.  Then each thread can have its private variable.  We can do the resize in TinyProfiler::Initiallize that is already called by amrex::Initialize.

**Review by WeiqunZhang — CHANGES_REQUESTED** [review]

File: Src/Base/AMReX_CArena.cpp (hunk 1 above)
Comment (AlexanderSinn) [reply Src/Base/AMReX_CArena.cpp]: Yes

File: Src/Base/AMReX_TinyProfiler.H (hunk 2 above)
Comment (AlexanderSinn) [reply Src/Base/AMReX_TinyProfiler.H]: I added this. That approach is prone to false sharing, so I made sure the distance between thread private objects is large enough.

Comment (WeiqunZhang) [reply Src/Base/AMReX_TinyProfiler.H]: You can have vector or unique_ptr of the objects.  That probably will be fine.

**Comment by WeiqunZhang** [comment]
I find the concept of AvgMem confusing.  For example, how should we think about `43 KiB`?
```
---------------------------------------------------------------------------
Name                                      Nalloc  Nfree    AvgMem    MaxMem
---------------------------------------------------------------------------
The_Arena::Initialize()                        1      1    43 KiB    29 GiB
```

**Comment by AlexanderSinn** [comment]
```
---------------------------------------------------------------------------
Name                                      Nalloc  Nfree    AvgMem    MaxMem
---------------------------------------------------------------------------
The_Arena::Initialize()                        1      1    43 KiB    29 GiB
```
Here AvgMem is significantly less than MaxMem. Functions like this require memory when they execute, but give it back to the system shortly after. In this case it is deallocated immediately. If the memory footprint of such a function where to be reduced, it may help the application as a whole very little unless it would run out of memory in this function.
```
---------------------------------------------------------------------------
Name                                      Nalloc  Nfree    AvgMem    MaxMem
---------------------------------------------------------------------------
Fields::AllocData()                            3      3  2820 MiB  2820 MiB
```
Here AvgMem is the same as MexMem. This function allocates a chunk of memory at the start and keeps it for the lifetime of the program. Reducing the amount of memory used in this function would allow the entire rest of the program to use more memory.
```
---------------------------------------------------------------------------
Name                                      Nalloc  Nfree    AvgMem    MaxMem
---------------------------------------------------------------------------
PlasmaParticleContainer::InitParticles        34     34  4554 MiB  4737 MiB
```
Here AvgMem is a little less than MaxMem. In this case it is caused by allocating a chunk of memory and keeping it until the program ends, however additional memory is required for initialization that if freed early. It may make more sense to try to reduce AvgMem rather than MaxMem, unless MaxMem is too large.

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**